File: algorithms/haskell/Knapsack.hs
import Data.Array.Unboxed
import Data.Time.Clock
import System.IO
import Control.Exception (evaluate)

knapsack :: Int -> [(Int,Int)] -> Int
knapsack cap items =
    let n   = length items
        arr = listArray (1,n) items :: Array Int (Int,Int)
        dp  = array ((0,0),(n,cap))
                    [((i,w), cell i w) | i <- [0..n], w <- [0..cap]]
        cell 0 _ = 0
        cell i w =
            let (wi, vi) = arr ! i
                skip     = dp ! (i-1, w)
            in if wi > w then skip
               else max skip (dp ! (i-1, w-wi) + vi)
    in dp ! (n, cap)

main :: IO ()
main = do
    hSetBuffering stdout LineBuffering
    contents <- getContents
    let ls      = lines contents
        first   = words (head ls)
        cap     = read (head first)  :: Int
        n       = read (first !! 1)  :: Int
        items   = map (\l -> let [w,v] = map read (words l)
                              in (w,v)) (take n (tail ls)) :: [(Int,Int)]

    start  <- getCurrentTime
    result <- evaluate (knapsack cap items)
    end    <- getCurrentTime

    let ms = realToFrac (diffUTCTime end start) * 1000 :: Double
    print result
    hPutStrLn stderr $ show ms

File: algorithms/lua/knapsack.lua
local function knapsack(W, weights, values, n)
    local dp = {}
    for w = 0, W do dp[w] = 0 end

    for i = 1, n do
        local wi, vi = weights[i], values[i]
        for w = W, wi, -1 do
            local candidate = dp[w - wi] + vi
            if candidate > dp[w] then dp[w] = candidate end
        end
    end
    return dp[W]
end

local function main()
    local first = io.read("*l")
    local W, n  = first:match("(%d+)%s+(%d+)")
    W = tonumber(W); n = tonumber(n)

    local weights = {}
    local values  = {}
    for i = 1, n do
        local line   = io.read("*l")
        local wi, vi = line:match("(%d+)%s+(%d+)")
        weights[i]   = tonumber(wi)
        values[i]    = tonumber(vi)
    end

    local start  = os.clock()
    local result = knapsack(W, weights, values, n)
    local ms     = (os.clock() - start) * 1000

    print(result)
    io.stderr:write(string.format("%.3f\n", ms))
end

main()